test/unit/TestLanguagePipeline.hs

```diff
@@ -417,3 +411,14 @@ testLanguagePipeline =
 
   getSyntax :: ProcessedTerm -> Syntax' Polytype
   getSyntax (ProcessedTerm (Module s _) _ _) = s
+
+-- | Check round tripping of term from and to text, then test ToJSON/FromJSON.
+roundTripTerm :: Text -> Assertion
+roundTripTerm txt = do
+  assertEqual "roundtrip (readTerm -> prettyText)" txt (prettyText term)
+  assertEqual "roundtrip (ToJSON -> FromJSON)" term (decodeThrow $ encode term)
+ where
+  decodeThrow v = case eitherDecode v of
+    Left e -> error $ "Decoding of " <> from (T.decodeUtf8 (from v)) <> " failed with: " <> from e
+    Right x -> x
+  term = fromMaybe (error "empty document") $ either (error . T.unpack) id $ readTerm txt
```

Comment:
@byorgey I changed the function to produce better errors on round tripping and not elaborate.
```diff
- roundtrip
- expected: ProcessedTerm (Module {moduleAST = Syntax' {_sLoc = SrcLoc 0 30, _sTerm = SBind Nothing (Syntax' {_sLoc = SrcLoc 0 14, _sTerm = SDef False (LV {lvSrcLoc = SrcLoc 4 6, lvVar = "x"}) Nothing (Syntax' {_sLoc = SrcLoc 8 11, _sTerm = TInt 41, _sType = Forall [] (TyBaseF BInt)}), _sType = Forall [] (TyCmdF (TyBaseF BUnit))}) (Syntax' {_sLoc = SrcLoc 16 30, _sTerm = SDef False (LV {lvSrcLoc = SrcLoc 20 22, lvVar = "y"}) Nothing (Syntax' {_sLoc = SrcLoc 24 27, _sTerm = TInt 42, _sType = Forall [] (TyBaseF BInt)}), _sType = Forall [] (TyCmdF (TyBaseF BUnit))}), _sType = Forall [] (TyCmdF (TyBaseF BUnit))}, moduleCtx = Ctx {unCtx = fromList [("x",Forall [] (TyBaseF BInt)),("y",Forall [] (TyBaseF BInt))]}}) (Requirements {capReqs = fromList [CPower,CEnv], devReqs = fromList [], invReqs = fromList []}) (Ctx {unCtx = fromList [("x",Requirements {capReqs = fromList [], devReqs = fromList [], invReqs = fromList []}),("y",Requirements {capReqs = fromList [], devReqs = fromList [], invReqs = fromList []})]})
- but got: ProcessedTerm (Module {moduleAST = Syntax' {_sLoc = SrcLoc 0 32, _sTerm = SBind Nothing (Syntax' {_sLoc = SrcLoc 0 15, _sTerm = SDef False (LV {lvSrcLoc = SrcLoc 4 6, lvVar = "x"}) Nothing (Syntax' {_sLoc = SrcLoc 9 12, _sTerm = TInt 41, _sType = Forall [] (TyBaseF BInt)}), _sType = Forall [] (TyCmdF (TyBaseF BUnit))}) (Syntax' {_sLoc = SrcLoc 17 32, _sTerm = SDef False (LV {lvSrcLoc = SrcLoc 21 23, lvVar = "y"}) Nothing (Syntax' {_sLoc = SrcLoc 26 29, _sTerm = TInt 42, _sType = Forall [] (TyBaseF BInt)}), _sType = Forall [] (TyCmdF (TyBaseF BUnit))}), _sType = Forall [] (TyCmdF (TyBaseF BUnit))}, moduleCtx = Ctx {unCtx = fromList [("x",Forall [] (TyBaseF BInt)),("y",Forall [] (TyBaseF BInt))]}}) (Requirements {capReqs = fromList [CPower,CEnv], devReqs = fromList [], invReqs = fromList []}) (Ctx {unCtx = fromList [("x",Requirements {capReqs = fromList [], devReqs = fromList [], invReqs = fromList []}),("y",Requirements {capReqs = fromList [], devReqs = fromList [], invReqs = fromList []})]})
+ roundtrip (readTerm -> prettyText)
+ expected: "def x = 41 end; def y = 42 end"
+ but got: "def x = 41 end;\ndef y = 42 end"
```

AFAIK `readTerm` will not typecheck, is that OK or should I use another function?

Reply:
I would like to eventually use this function for #1469, thought that might mean either a copy in integration tests or a common test utility library.

Reply:
Yes, it makes sense to use `readTerm` here instead of `processTerm`.